ndb/src/google/cloud/ndb/_eventloop.py

```diff
@@ -259,6 +263,59 @@ def run(self):
                 break
 
 
+class _LocalContexts(threading.local):
+    """Maintain a thread local stack of event loops."""
+
+    def __init__(self):
+        self.stack = []
+
+    def push(self, loop):
+        self.stack.append(loop)
+
+    def pop(self):
+        return self.stack.pop(-1)
+
+    def current(self):
+        if self.stack:
+            return self.stack[-1]
+
+
+contexts = _LocalContexts()
+
+
+@contextlib.contextmanager
+def async_context():
+    """Establish a context for a set of asynchronous API calls.
+
+    This function provides a context manager which establishes the event loop
+    that will be used for any asynchronous NDB calls that occur in the context.
+    For example::
+
+        from google.cloud.ndb import async_context
+
+        with async_context():
+            # Make some asynchronous calls
+            pass
+
+    Within the context, any calls to a ``*_async`` function or to an
+    ``ndb.tasklet``, will be added to the event loop established by the
+    context.  Upon exiting the context, execution will block until all
+    asynchronous calls loaded onto the event loop have finished execution.
+
+    Code within an ``async_context`` should be single threaded. Internally, a
+    ``threading.local`` instance is used to track the current event loop.
+
+    In the context of a web application, it is recommended that a single
+    ``async_context`` be used per HTTP request. This can typically be
+    accomplished in a middleware layer.
+    """
+    loop = EventLoop()
+    contexts.push(loop)
+    yield
+    contexts.pop()
```

Comment:
@chrisrossi Re-starting the conversation here about whether `loop` is being popped / what to do if another `EventLoop` instance comes off instead

Reply:
I don't think that can happen unless someone is accessing our private data structure without our permission. I can throw an `assert` in there, just in case.

Reply:
Its a thread local, and documented that user code is supposed to be single-threaded.  The only way that would happend is if the user code pushed a loop onto the context manually, which would just be the Wrong Thing(TM), no?

Reply:
Indeed. So long as `ndb._eventloop.contexts` is non-public, we need not worry.

Rather than an `assert` how about just a comment about the assumption we're making?